Update WKDDActionContext serialization to use WKSecureCoding interface

sheeparegreat opened the pull request:
#### 4a8681f07b140500eba88271dd9c645d88c11416
<pre>
Update WKDDActionContext serialization to use WKSecureCoding interface
<a href="https://bugs.webkit.org/show_bug.cgi?id=301588">https://bugs.webkit.org/show_bug.cgi?id=301588</a>
<a href="https://rdar.apple.com/163590357">rdar://163590357</a>

Reviewed by Alex Christensen.

This patch removes WKKeyedCoder at the IDL layer for serializing WKDDActionContext.

IPCSerialization.mm tests are refactored to split out the Data Detectors test.

Test: Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::typeFromObject):
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
(WebKit::valueFromID):
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/mac/CoreIPCDDSecureActionContext.h: Added.
* Source/WebKit/Shared/mac/CoreIPCDDSecureActionContext.mm: Added.
(WebKit::CoreIPCDDSecureActionContext::CoreIPCDDSecureActionContext):
(WebKit::CoreIPCDDSecureActionContext::toID const):
* Source/WebKit/Shared/mac/CoreIPCDDSecureActionContext.serialization.in: Renamed from Source/WebKit/Shared/Cocoa/CoreIPCDDActionContext.serialization.in.
* Source/WebKit/Shared/mac/WebHitTestResultPlatformData.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST(IPCSerialization, DataDetectors)):
(TEST(IPCSerialization, SecureCoding)):

Canonical link: <a href="https://commits.webkit.org/302731@main">https://commits.webkit.org/302731@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c54909b0bca998bdc7da3d55fb9aa5ebe3828591

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129937 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2198 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40795 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137330 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81432 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2cb3a0c7-b609-4d65-a062-7713f105176e) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131808 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2235 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2089 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98978 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66787 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7a0c0ae1-fa8e-44ac-85f4-27eeca740aa4) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132884 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1661 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116371 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79671 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0a17bed4-b834-4f14-89b9-989ca0494ba8) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1580 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34498 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80600 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110105 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35007 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139811 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1992 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1921 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107483 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2037 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112721 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107371 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27351 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1593 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31190 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54794 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2065 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65434 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1880 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1914 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1988 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sheeparegreat commented on Source/WebKit/Shared/Cocoa/CoreIPCDDActionContext.serialization.in:
```diff
@@ -20,35 +20,4 @@
 # OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-#if PLATFORM(MAC) && HAVE(WK_SECURE_CODING_DATA_DETECTORS)
```
Remove this file

sheeparegreat commented on Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
```diff
@@ -76,7 +76,11 @@ static ObjectValue valueFromID(id object)
 #if PLATFORM(MAC)
     case IPC::NSType::WKDDActionContext:
-        return CoreIPCDDSecureActionContext((WKDDActionContext *)object);
+        // Not serialized via CoreIPCNSCFObject
```
Change this to completely remove. Use a new member on WKKeyedCoder to enable this

achristensen07 commented on Source/WebKit/Shared/mac/CoreIPCDDSecureActionContext.h:
```diff
@@ -0,0 +1,85 @@
+    std::optional<RetainPtr<DDScannerResult>> mainResult;
+
+    RetainPtr<NSNumber> immediate;
+    RetainPtr<NSNumber> isRightClick;
```
This and immediate seem like they could be just a bool

achristensen07 commented on Source/WebKit/Shared/mac/CoreIPCDDSecureActionContext.mm:
```diff
@@ -0,0 +1,226 @@
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+    NSValue *highlightFrame = [dictionary.get() objectForKey:@"highlightFrame"];
+    if ([highlightFrame isKindOfClass:[NSValue class]]) {
+        if (!strcmp(highlightFrame.objCType, @encode(NSRect))) { // NOLINT
```
Other places where we use objCType, we do something like this:
NSString *type = [NSString stringWithCString:[valueObject objCType] encoding:NSASCIIStringEncoding];
That would probably be better than strcmp and WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN/END

achristensen07 commented on Source/WebKit/Shared/mac/CoreIPCDDSecureActionContext.mm:
```diff
@@ -0,0 +1,226 @@
+        m_data.matchedString = matchedString;
+
+    NSArray *allResults = [dictionary.get() objectForKey:@"allResults"];
+    if ([allResults isKindOfClass:NSArray.class]) {
```
dynamic_objc_cast would make this a bit cleaner:
if (NSArray *allResults = dynamic_objc_cast<NSArray>([dictionary objectForKey:@"allResults"]))

sheeparegreat committed:
Update WKDDActionContext serialization to use WKSecureCoding interface
https://bugs.webkit.org/show_bug.cgi?id=301588
rdar://163590357

Reviewed by Alex Christensen.

This patch removes WKKeyedCoder at the IDL layer for serializing WKDDActionContext.

IPCSerialization.mm tests are refactored to split out the Data Detectors test.

Test: Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::typeFromObject):
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
(WebKit::valueFromID):
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/mac/CoreIPCDDSecureActionContext.h: Added.
* Source/WebKit/Shared/mac/CoreIPCDDSecureActionContext.mm: Added.
(WebKit::CoreIPCDDSecureActionContext::CoreIPCDDSecureActionContext):
(WebKit::CoreIPCDDSecureActionContext::toID const):
* Source/WebKit/Shared/mac/CoreIPCDDSecureActionContext.serialization.in: Renamed from Source/WebKit/Shared/Cocoa/CoreIPCDDActionContext.serialization.in.
* Source/WebKit/Shared/mac/WebHitTestResultPlatformData.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST(IPCSerialization, DataDetectors)):
(TEST(IPCSerialization, SecureCoding)):

Canonical link: https://commits.webkit.org/302731@main
